### PR TITLE
Focus stack frame if DA issues InvalidatedEvent for stack

### PR DESCRIPTION
**Pull request by AHodkinsonIMG**
This is the first pass at a fix for https://github.com/microsoft/vscode/issues/172544 following the suggested method.  
I'm not familiar with the VSCode codebase and the debug workbench so please go easy if what I've written is garbage!

This can be tested with https://github.com/AHodkinsonIMG/vscode-mock-debug, the procedure for testing the bug is written up in the issue. 

In the case that the InvalidatedEvent doesn't contain a threadId I have chosen to skip over the stack frame update.  An alternative might be to say that if there is no threadId then we should update _all_ available threads?

### CODE REVIEW

**Comment by AHodkinsonIMG** [comment]
@microsoft-github-policy-service agree company="Imagination Technologies"

File: src/vs/workbench/contrib/debug/browser/debugSession.ts [hunk 1]
```diff
@@ -1245,6 +1245,30 @@ export class DebugSession implements IDebugSession {
...
+								// Only take focus if nothing is focused, or if the focus is already on the current session
+								const preserveFocus = !this.configurationService.getValue<IDebugConfiguration>('debug').focusEditorOnBreak;
+								await this.debugService.focusStackFrame(undefined, thread, undefined, { preserveFocus });
+							}
+						}
+					};
+					await promises.topCallStack;
+					focus();
+					await promises.wholeCallStack;
+					focus();
```
Comment (roblourens) [inline src/vs/workbench/contrib/debug/browser/debugSession.ts:1270]: Are you sure that calling focus() twice is right?

Comment (AHodkinsonIMG) [reply src/vs/workbench/contrib/debug/browser/debugSession.ts:1270]: The code in the stop event handler (https://github.com/microsoft/vscode/blob/41f18661e4af20b9ac04fb06f0d8c70ea072ac85/src/vs/workbench/contrib/debug/browser/debugSession.ts#L979) has a similar construct, and this seems to be used to keep the UI performant in the case of a large stack which may take a long-time to read from the debug adapter.
If the InvalidatedEvent is received for the 'stack' we don't know which part of the stack has changed so we need to read the entire stack again, and could also suffer from this issue?

**Comment by roblourens** [comment]
Sorry for taking so long on this. Finally had the chance to revisit it. In the meantime I even looked at the same issue filed by @sanannavyaa and forgot that we already had a PR to fix it 😅. 

I think that in that scenario, there is no thread id. Like you point out, I think that we should update all threads in that case. We already fetch all threads anyway. Does that make sense for your scenario @sanannavyaa? It looks like you don't sent a thread id.

In the other issue, I suggested that it might work better to send another StoppedEvent instead of using InvalidatedEvent. That is probably still true but I'm fine with this approach too.

**Review by sanannavyaa — APPROVED** [review]
@roblourens 

I tested this out and it does not seem to fix the issue on our side. My guess is that because this PR seeks to specifically fix focusing on the stack when InvalidatedEvent['stack'] is issued, it does not help us since we use InvalidatedEvent['all']. I suspect that this is part of the solution but not the full solution that we would need to set the line highlight focus correctly after we issue InvalidatedEvent['all'].

For the issue on our side, it's still the same, we change focus, send InvalidatedEvent['all'] and the line highlight does not update correctly in the UI even though we see the correct line number in the debug console. I am attaching a screenshot for greater detail. I request we reopen https://github.com/microsoft/vscode/issues/189691. Just a side note - the alternate option where you recommended that we issue a StoppedEvent works for us. 

That said, for the issue it seeks to resolve, I think the commit is good so I am approving it.
InvalidatedEventBug

**Comment by roblourens** [comment]
I know it doesn't fix your scenario as-is, I suggested that we would update all threads when no thread id is passed, which I think would fix your scenario as well.

**Comment by sanannavyaa** [comment]
That sounds great. I agree that should fix it and I am happy to test it once that is up as a PR or help in any other way I can. Thanks!

**Comment by AHodkinsonIMG** [comment]
@roblourens I'm happy to add that change in ot my PR.
To check my understanding of the fix:
* if no thread id is specified in the DAP InvalidatedEvent then we'd use `this.getAllThreads()` to get all the threads for the debug session.
* for each of these threads we refresh the callstack using `this.model.refreshTopOfCallstack(thread)`.  Is there any advantage to running these updates in parallel, I'm assuming that the DA will be responding in a serial manner so maybe not?

**Comment by roblourens** [comment]
> I'm assuming that the DA will be responding in a serial manner so maybe not?

It may or may not- I think the right way to handle this from the vscode side is in a parallel way.

**Comment by connor4312** [comment]
I didn't realize this was a PR, I may have supplanted it in https://github.com/microsoft/vscode/pull/200789. In that PR, when a stack invalidation happens, I rerun logic as if the `stopped` event was newly received.

**Comment by roblourens** [comment]
I believe this is actually covered now

**Pull request closed without merging**